articles/external_qos_file.md

```diff
@@ -0,0 +1,658 @@
+---
+# Current response definition
+...
+```
+
+A `rcl_qos_loaded_profiles_t` can be constructed from the received message and passed to the node options of the node to be loaded.
+
+TBD: QoS message definition.
+
+### QoS profiles for actions
+
+Currently, there is a collection of [5 QoS profiles](https://github.com/ros2/rcl/blob/af438bcca1ffb69714c0b225ba0f80aceab04906/rcl_action/include/rcl_action/action_client.h#L44-L52) to be set for creating an action, that will set the QoS profiles of the topics and services making up the action.
+The file format should support actions, if not the user would have to use the topics and service names with the mangling `rcl` applies, which is a bit hard to remember.
+For the most common use cases, it makes sense to set the 5 QoS profiles to the same settings:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <action_client name="my_action">
+            <qos>
+                <history_depth>100</history_depth>
+            </qos>
+        </action_client>
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        action:
+            name: my_action
+            qos:
+                history_depth: 100
+```
+
+The file format could also support to set each of the QoS profiles individually:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <action_client name="my_action">
+            <feedback>
+                <qos>
+                    <history_depth>100</history_depth>
+                </qos>
+            <feedback>
+        </action_client>
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        action_client:
+            name: my_action
+            feedback:
+                qos:
+                    history_depth: 100
+```
+
+It would be worth revisiting if having 5 different QoS profiles for defining actions make sense before extending the file format in such a way.
+
+### Support in launch files
+
+`launch_ros.actions.Node` and `launch_ros.actions.LoadComposableNode` actions could support a qos file argument in their constructors.
+
+### rmw implementation specific payload
+
+There are currently QoS settings that are supported by DDS and not by ROS.
+An user might want to make use of them, and they won't find an easy way to do that without exposing the QoS setting on ROS.
+
+There are some cases where an user might want to use a vendor specific QoS extension, for example rti allows defining [flow controllers](https://community.rti.com/static/documentation/connext-dds/5.2.0/doc/manuals/connext_dds/html_files/RTI_ConnextDDS_CoreLibraries_UsersManual/Content/UsersManual/CreatingandConfigCustomFlowC.htm) in publishers.
+
+Though a DDS vendor specific QoS profile file can be used to set up these QoS settings, you currently can only use the default qos profile.
```

Comment:
> I don't quite understand this point. I think you are you saying that we can only specify the default QoS for the rmw vendor, and not customize the settings per node entity. Is that right?

You can only set a profile called "default", that's it.
Every entity will use that profile (publishers, subscriptions, etc).

Because of how the rmw implementation is coded, the policies we're already using won't be read from that profile, except `RMW_QOS_POLICY_*_SYSTEM_DEFAULT` was used. For the policies we're not using (e.g. flow controllers), those would be applied to all entities.

Reply:
Rephrased in https://github.com/ros2/design/pull/296/commits/75df29cf26d057598c0c76bd14b77c64c6317b74.